core/store/src/trie/mem/arena/concurrent.rs

```diff
@@ -0,0 +1,252 @@
+use super::alloc::{allocation_class, allocation_size, CHUNK_SIZE};
+use super::{Arena, ArenaMemory, ArenaPos, ArenaSliceMut, STArena};
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::sync::Arc;
+
+/// Arena that can be allocated on from multiple threads, but still allowing conversion to a
+/// single-threaded `STArena` afterwards.
+///
+/// The `ConcurrentArena` cannot be directly used; rather, for each thread wishing to use it,
+/// `for_thread` must be called to get a `ConcurrentArenaForThread`, which then acts like a
+/// normal arena, except that deallocation is not supported.
+///
+/// The only synchronization they need is the chunk counter. The purpose is so that after
+/// multiple threads allocate on their own arenas, the resulting memory can still be combined
+/// into a single arena while allowing the pointers (ArenaPos) to still be valid. This is what
+/// allows a memtrie to be loaded in parallel.
+pub struct ConcurrentArena {
+    /// Chunks allocated by
+    next_chunk_pos: Arc<AtomicUsize>,
+}
+
+impl ConcurrentArena {
+    pub fn new() -> Self {
+        Self { next_chunk_pos: Arc::new(AtomicUsize::new(0)) }
+    }
+
+    /// Returns an arena that can be used for one thread.
+    pub fn for_thread(&self) -> ConcurrentArenaForThread {
+        ConcurrentArenaForThread::new(self.next_chunk_pos.clone())
+    }
+
+    /// Converts the arena to a single-threaded arena. All returned values of `for_thread` must be
+    /// passed in.
+    ///
+    /// There is a caveat that may be fixed in the future if desired: the returned arena will have
+    /// some memory wasted. This is because the last chunk of each thread may not be full, and
+    /// the single-threaded arena is unable to make use of multiple partially filled chunks. The
+    /// maximum memory wasted is 4MB * number of threads; the average is 2MB * number of threads.
+    /// The wasted memory will be reclaimed when the memtrie shard is unloaded.
+    pub fn to_single_threaded(
+        self,
+        name: String,
+        threads: Vec<ConcurrentArenaForThread>,
+    ) -> STArena {
+        let mut chunks = Vec::new();
+        let mut active_allocs_bytes = 0;
+        let mut active_allocs_count = 0;
+        for thread in threads {
+            let memory = thread.memory;
+            let mut chunk_pos = vec![usize::MAX; memory.chunks.len()];
+            for (pos, index) in memory.chunk_pos_to_index.into_iter().enumerate() {
+                if index == usize::MAX {
+                    continue;
+                }
```

Comment:
Assertion `chunk_pos[index] == usize::MAX` also seems to make sense here.

Reply:
Now that I read this code again, it is needlessly complicated that I myself can't understand it.

I fixed that by just storing the global chunk pos in the list of local chunks in `ConcurrentArenaForThread`. It makes the whole thing much simpler and also fixes a bug that the previous assertion for "not all arena threads have been passed in" was not actually asserting the right thing.